### ISSUE TITLE
rundeck job without network activity during 2 hours never ends and are fail

### ISSUE THREAD

**Issue by philjov** [rundeck/rundeck#5617]
**Describe the bug**
A job with 3 steps
Node is Oracle Linux Server release 7.7
script is perl with rsync
script works when scheduled in crontab
there is no timeout set on the job
**My Rundeck detail**
Nothing in logs, even with debug 
* Rundeck version: [ e.g. 2.11.3]
3.1.2
* install type: [choose: rpm,deb,war,launcher]
rpm
* OS Name/version: [e.g. ubuntu 16.04]
Oracle Linux Server release 7.7
* DB Type/version: [choose: mysql, postgres, mssql, oracle, ... ]
PostGreSQL 9.6.11 in AWS RDS

**To Reproduce**
Run a job with "sleep 7210" 
After 2 hours (7200 seconds)
- While checking in the Linux server no process there
- Rundeck is hanging indefinitely waiting something.
- SSH channel between rundeck and node is broken, then all process on node are stopped, but rundeck don't see anything.

**Expected behavior**
Process continue after 7200 seconds and job finish normally.

**Screenshots**
N/A
**Desktop (please complete the following information):**
N/A
**Additional context**
N/A

**Comment by MegaDrive68k** [comment]
Hi @philjov 

Any clue if you check the `service.log` file on the execution moment? the file is located at `/var/log/rundeck/service.log`

You can check it on "real time" with `tail -f /var/log/rundeck/service.log`.

Kind regards.

**Comment by philjov** [comment]
Hello, I updated bug report.

There is nothing in /var/log/rundeck/service.log related to the problem.

But I analyzed the job and seems that nothing happen between 2 Steps, and it could be several hours between 2 Steps, so I checked the SSH keep alive between rundeck server and the node, it was configured by default. 
In order to test this, on rundeck server I added "ServerAliveInterval 120" in  /etc/ssh/ssh_config (openssh-clients-7.4p1-21.el7.x86_64)
and on node I added following lines
ClientAliveInterval 120
ClientAliveCountMax 5040
in /etc/ssh/sshd_config  (openssh-server-7.4p1-21.el7.x86_64)
and sshd service was restarted on node.
Because the question for me is: Is rundeck perform a keepalive in order to keep ssh channel open even if nothing happen during hours.

Thanks a lot for help.

**Comment by MegaDrive68k** [comment]
The best way to configure that is in this way (if you use "SSH node executor" the default Rundeck SSH node executor):

https://docs.rundeck.com/docs/administration/projects/node-execution/ssh.html#specifying-ssh-timeout-options

Test with that. Those parameters work for Jsch library, the default SSH library for Rundeck (http://www.jcraft.com/jsch/).  

Also, you can set your node executor as "OpenSSH" and your configuration should be work.

Hope it helps!

**Comment by philjov** [comment]
Hello, 

thanks, but I didn't found any timeout in rundeck:
project level: Connection Timeout 0 Command Timeout 0
framework.properties:framework.ssh.timeout = 0

but this timeout seems to be the good one:
cat /proc/sys/net/ipv4/tcp_keepalive_time
7200

In fact I monitored the process activity in the rundeck server, and I ran a tcpdump

Here's the result:
- Processes are ending after 2 hours of activity
- There is no network activity during this time. (At the beginning and after)

Rundeck have a keepalive  problem on linux servers.

If the job task run rometely on the node without network activity the tcp_keepalive will close the tcp socket and the linux server kill the processes which are below it.
And rundeck don't see any problem and wait indefinetly a feedback.

Problem 1: Rundeck should manage a keepalive activity for jobs which have no network activity.

Problem 2: Rundeck is waiting a feedback although the network socket is closed and no processes are running in the node anymore.

Best regards,

Philippe
